File: src/Oisys.Web/Models/OrderLineItem.cs
﻿namespace OisysNew.Models
{
    /// <summary>
    /// <see cref="OrderLineItem"/> class represents the child of Order object.
    /// </summary>
    public class OrderLineItem : ModelBase
    {
        /// <summary>
        /// Gets or sets property OrderId.
        /// </summary>
        public long OrderId { get; set; }

        /// <summary>
        /// Gets or sets property Quanity.
        /// </summary>
        public int Quantity { get; set; }

        /// <summary>
        /// Gets or sets property ItemId.
        /// </summary>
        public long ItemId { get; set; }

        /// <summary>
        /// Gets or sets the unit price for the item.
        /// </summary>
        public decimal UnitPrice { get; set; }

        /// <summary>
        /// Gets or sets the discounted unit price.
        /// </summary>
        public decimal DiscountedUnitPrice { get; set; }

        /// <summary>
        /// Gets or sets quantity returned for this line item.
        /// </summary>
        public int QuantityReturned { get; set; }

        /// <summary>
        /// Gets or sets quantity delivered for this line item.
        /// </summary>
        public int QuantityDelivered { get; set; }

        /// <summary>
        /// Gets or sets quantity invoiced for order line item.
        /// </summary>
        public int QuantityInvoiced { get; set; }

        /// <summary>
        /// Gets or sets Order navigation property.
        /// </summary>
        public virtual Order Order { get; set; }

        /// <summary>
        /// Gets or sets Item navigation property.
        /// </summary>
        public virtual Item Item { get; set; }

        /// <summary>
        /// Gets or sets the transaction history navigation property.
        /// </summary>
        public virtual ItemHistory TransactionHistory { get; set; }
    }
}

File: src/Oisys.Web/ClientApp/src/app/shared/models/order-line-item.ts
export class OrderLineItem {
  public id: number;
  public orderCode: string;
  public orderDate: string;
  public itemId: number;
  public itemCode: string;
  public itemName: string;
  public quantity: number;
  public unit: string;
  public unitPrice: number;
  public categoryName: string;
  public quantityReturned: number;
  public quantityDelivered: number;
  public discountedUnitPrice: number;

  get totalPrice() {
    return this.quantity * this.unitPrice;
  }

  get quantityNotDelivered() {
    return this.quantity - this.quantityDelivered;
  }

  constructor();
  constructor(orderLineItem: OrderLineItem);
  constructor(orderLineItem?: any) {
    this.id = orderLineItem && orderLineItem.id || 0;
    this.orderCode = orderLineItem && orderLineItem.orderCode || '';
    this.orderDate = orderLineItem && orderLineItem.orderDate || '';
    this.itemId = orderLineItem && orderLineItem.itemId || 0;
    this.itemCode = orderLineItem && orderLineItem.itemCode || '';
    this.itemName = orderLineItem && orderLineItem.itemName || '';
    this.quantity = orderLineItem && orderLineItem.quantity || 0;
    this.unit = orderLineItem && orderLineItem.unit || '';
    this.unitPrice = orderLineItem && orderLineItem.unitPrice || 0;
    this.categoryName = orderLineItem && orderLineItem.categoryName || '';
    this.quantityReturned = orderLineItem && orderLineItem.quantityReturned || 0;
    this.quantityDelivered = orderLineItem && orderLineItem.quantityDelivered || 0;
    this.discountedUnitPrice = orderLineItem && orderLineItem.discountedUnitPrice || 0;
  }
}
